src/lavinmq/amqp/stream/stream_message_store.cr

```diff
@@ -86,8 +80,15 @@ module LavinMQ::AMQP
     private def offset_at(seg, pos, retried = false) : Tuple(Int64, UInt32, UInt32)
       return {@last_offset, seg, pos} if @size.zero?
       mfile = @segments[seg]
-      msg = BytesMessage.from_bytes(mfile.to_slice + pos)
-      offset = offset_from_headers(msg.properties.headers)
+      offset = @offset_index[seg]
+      mfile.pos = 4
+      loop do
+        BytesMessage.skip(mfile)
+        break if mfile.pos >= pos
```

Comment:
no need to read the last message (eg. looking for the offset of message as `pos` 4). 
```suggestion
      while mfile.pos < pos
        BytesMessage.skip(mfile)
```